### PR TITLE
Fix Problem with substitution and paranthesis handling

### LINKED ISSUE #13333: New problem with substitution and paranthesis handling
Something has changed in the way substitutions work in sympy version 1.1.1. on python3.4.2

```
In [3]: sympy.S('R*F').subs({'R':'8314.4*mjoule/(mole*kelvin)','F':'96484.6*(coulomb/mole)'})
Out[3]: 8314.4*mjoule/(mole*kelvin)*96484.6*(coulomb/mole)

In [4]: sympy.S('R/F').subs({'R':'8314.4*mjoule/(mole*kelvin)','F':'96484.6*(coulomb/mole)'})
Out[4]: 8314.4*mjoule/(mole*kelvin)/96484.6*(coulomb/mole)

```

with sympy 1.0 on python2.7.9 this yielded

```
In [3]: sympy.S('R*F').subs({'R':'8314.4*mjoule/(mole*kelvin)','F':'96484.6*(coulomb/mole)'})
Out[3]: 802211558.24*coulomb*mjoule/(kelvin*mole**2)

In [4]: sympy.S('R/F').subs({'R':'8314.4*mjoule/(mole*kelvin)','F':'96484.6*(coulomb/mole)'})
Out[4]: 0.0861733375067109*mjoule/(coulomb*kelvin)
```

The update seems to have wronged something.

### LINKED ISSUE #13654: Substitution lacks parenthesis
The following inconsistency in substitutions has emerged in `sympy 1.1.1` and is different to the behaviour found in `sympy 1.0`: 

```
fct = sympy.S("1/f")
A = {"f": "(1 / 2)"}
B = {"f": "1 / 2"}
print fct.subs(A)
print fct.subs(B)
print sympy.simplify(fct.subs(A) - fct.subs(B)) == 0
```

which gives

```
1/(1 / 2)  [= 2 as expected]
1/1 / 2    [= 1/2 which is wrong]
False      [confirming that both expressions are not equal, as they should be]
```

I think this could be solved by adding - by default - a parenthesis around expressions before substituting.

This issue has already been reported in [https://github.com/sympy/sympy/issues/13333](https://github.com/sympy/sympy/issues/13333), which filed a slightly more difficult example.

### PR DESCRIPTION
**Pull request by Subhash-Saurabh**
Fixes #13654 
Fixes #13333

 While substitution  sympify was called with strict=True , so sympify didn't correctly evaluated the terms. Now sympify is called twice once with strict=True and again with strict=False to allow proper evaluation of terms. 
```
>>> import sympy
>>> fct = sympy.S("1/f")
>>> A = {"f": "(1 / 2)"}
>>> B = {"f": "1 / 2"}
>>> print fct.subs(A)
>>> print fct.subs(B)
```
Earlier the output was
```
1/(1 / 2)  
1/1 / 2 
```
New output 
```
2
2
```

### CODE REVIEW

**Comment by Subhash-Saurabh** [comment]
@asmeurer  can you please review my PR.

File: sympy/core/basic.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 from .compatibility import (iterable, Iterator, ordered,
     string_types, with_metaclass, zip_longest, range)
 from .singleton import S
+from .core import all_classes as sympy_classes
```
Comment (asmeurer) [inline sympy/core/basic.py]: This is deprecated and shouldn't be used any more.

**Comment by asmeurer** [comment]
The automatic sympification of strings is something that shouldn't be relied on. 

On the other hand, converting a string to Symbol in subs seems even worse. I think we should just remove that, instead of calling sympify twice.

**Comment by asmeurer** [comment]
The change originates from 527de067a1d. I guess the motivation was to allow `subs(dict(a=1))` (https://github.com/sympy/sympy/issues/4680). 

It's been around for awhile, but I still don't like it. If we do keep it, we should restrict it to things that can actually be keyword arguments.

**Comment by mstimberg** [comment]
In case it helps: bisecting with the above mentioned example and checking whether `sympy.simplify(fct.subs(A) - fct.subs(B)) == 0` points to db2b54bb as the "first bad commit".

**Comment by Subhash-Saurabh** [comment]
Yes @mstimberg I think not allowing the expression to be completely sympify was the reason for this error.
However running sympify completely  on all expressions might bring back the errors that were dealt in [`527de06`](https://github.com/sympy/sympy/pull/10574/commits/db2b54bb59efb90ea04ff95c71d215698ac3db64)  and [`527de06`](https://github.com/sympy/sympy/commit/527de067a1db5a962fa25e86972ad76c60d2dbeb).
Therefore I made a workaround to fix this. However I do believe that running sympify with strict=False on strings might bring some new unseen errors.

**Comment by Subhash-Saurabh** [comment]
@asmeurer I have replaced the deprecated  `all_classes` . What other changes do you suggest ?

**Comment by Subhash-Saurabh** [comment]
@gxyd can you please review this PR and suggest what additional changes are needed.

**Comment by asmeurer** [comment]
I'm still -0.5 on this. I don't like adding more automatic sympifying of strings.

**Comment by Subhash-Saurabh** [comment]
@asmeurer 
If we do something like 
```
try:
    si = sympify(si)
except:
    if type(si) is str:
        si = Symbol(si)
```
it gives us the desired output but it fails when
 ```subs(dict(a=1))```

Do we really need to allow `subs(dict(a=1))` to evaluate.
 Can't we restrict this format and mention it in the documentation.

**Pull request closed without merging**